cmd/git-go-patch/extract.go

```diff
@@ -70,15 +101,142 @@ var extract = subcommand{
+			for _, cmd := range cmds {
+				if after, found := stringutil.CutPrefix(cmd, patchNumberCommand); found {
+					num, err := strconv.Atoi(after)
+					if err != nil {
+						return fmt.Errorf("malformed patch number command arg %q in %#q: %w", after, path, err)
+					}
+					if num >= n {
+						n = num
+					} else {
+						return fmt.Errorf("patch number command arg %v too small, expected at least %v in %#q", num, n, path)
+					}
+				} else {
+					return fmt.Errorf("command %#q is not recognized in patch %#q", cmd, path)
+				}
+			}
+
+			if n > 9999 {
+				return fmt.Errorf("rearranged patch number %v exceeds max 4-digit int used by patch naming convention: %#q", n, path)
+			}
+
+			oldNum, after, found := cutFilenameNumber(path)
+			if !found {
+				return fmt.Errorf("no number prefix found in %#q", path)
+			}
+
+			if n != oldNum {
+				newName := fmt.Sprintf("%04v-%v", strconv.Itoa(n), after)
+				if err := renameUnlessExists(path, filepath.Join(tempPatchDir, newName)); err != nil {
+					return fmt.Errorf("unable to rename %#q: %w", path, err)
+				}
+			}
+			n++
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		// Delete all old patches so if any commit descriptions have been changed, we don't end up
+		// with two copies of those patch files with slightly different names.
+		if err := patch.WalkGoPatches(rootDir, func(path string) error {
+			return os.Remove(path)
+		}); err != nil {
+			return err
+		}
+
+		// Move all patch files from the temp dir to the final dir.
+		if err := patch.WalkPatches(tempPatchDir, func(path string) error {
+			filename := filepath.Base(path)
+			log.Printf("Moving patch to destination: %#q\n", filename)
+			return os.Rename(path, filepath.Join(patchDir, filename))
+		}); err != nil {
+			return err
+		}
+
+		log.Printf("Extracted patch files from %#q into %#q\n", goDir, patchDir)
+		return nil
 	},
 }
+
+// readPatchCommands reads the given patch file's header and returns all potential commands, with
+// commandPrefix trimmed off.
+func readPatchCommands(path string) ([]string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var cmds []string
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+		t := scanner.Text()
+		if t == "---" {
+			// Header is done: stop reading. Technically, "---" could occur inside the commit
+			// message, so we might be giving up early. But even "git format-patch" and "git am"
+			// don't round-trip "---" ("format-patch" doesn't escape it, "am" cuts off the message),
+			// so don't worry about it here.
+			break
+		}
+		if after, found := stringutil.CutPrefix(t, commandPrefix); found {
+			cmds = append(cmds, after)
+		}
+	}
+	return cmds, nil
+}
+
+func renameUnlessExists(path, newPath string) error {
+	_, err := os.Stat(newPath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return os.Rename(path, newPath)
```

Comment:
Moving over to Linux where my temp is set up differently, I see this only works on one partition. (And maybe more limitations--the doc notes that it's only really intended to work within a single directory.) This isn't general-purpose like `mv`. Changing the code to use copies.